stdsimd-verify/src/lib.rs

```diff
@@ -0,0 +1,248 @@
+    ($($t:tt)*) => (quote_spanned!(proc_macro2::Span::call_site(), $($t)*))
+}
+
+#[proc_macro]
+pub fn x86_functions(input: TokenStream) -> TokenStream {
+    let dir = Path::new(env!("CARGO_MANIFEST_DIR"));
+    let root = dir.parent().unwrap();
+    let root = root.join("coresimd/src/x86");
+
+    let mut files = Vec::new();
+    walk(&root, &mut files);
+
+    let mut functions = Vec::new();
+    for file in files {
+        for item in file.items {
+            match item {
+                syn::Item::Fn(f) => functions.push(f),
+                _ => {}
+            }
+        }
+    }
+
+    functions.retain(|f| {
+        match f.vis {
+            syn::Visibility::Public(_) => {}
+            _ => return false,
+        }
+        match f.unsafety {
+            syn::Unsafety::Unsafe(_) => {}
+            _ => return false,
+        }
+        f.attrs.iter()
+            .filter_map(|a| a.meta_item())
+            .any(|a| {
+                match a {
+                    syn::MetaItem::NameValue(i) => i.ident == "target_feature",
+                    _ => false,
+                }
+            })
+    });
+
+    let input = proc_macro2::TokenStream::from(input);
+
+    let functions = functions.iter()
+        .filter(|f| {
+            !f.ident.sym.as_str().starts_with("_mulx")
+        })
+        .map(|f| {
+            let name = f.ident;
+            // println!("{}", name);
+            let mut arguments = Vec::new();
+            for input in f.decl.inputs.iter().map(|s| s.into_item()) {
+                let ty = match *input {
+                    syn::FnArg::Captured(ref c) => &c.ty,
+                    _ => panic!("invalid argument on {}", name),
+                };
+                arguments.push(to_type(ty));
+            }
+            let ret = match f.decl.output {
+                syn::ReturnType::Default => my_quote! { None },
+                syn::ReturnType::Type(ref t, _) => {
+                    let ty = to_type(t);
+                    my_quote! { Some(#ty) }
+                }
+            };
+            let instrs = find_instrs(&f.attrs);
+            let target_feature = find_target_feature(f.ident, &f.attrs);
+            my_quote! {
+                Function {
+                    name: stringify!(#name),
+                    arguments: &[#(#arguments),*],
+                    ret: #ret,
+                    target_feature: #target_feature,
+                    instrs: &[#(stringify!(#instrs)),*],
+                }
+            }
+        })
+        .collect::<Vec<_>>();
+
+    let ret = my_quote! { #input: &[Function] = &[#(#functions),*]; };
+    // println!("{}", ret);
+    ret.into()
+}
+
+fn to_type(t: &syn::Type) -> Tokens {
+    match *t {
+        syn::Type::Path(ref p) => {
+            match extract_path_ident(&p.path).sym.as_str() {
+                "__m128i" => my_quote! { &I8x16 },
+                "__m256i" => my_quote! { &I8x32 },
+                "__m64" => my_quote! { &I8x8 },
```

Comment:
This is a bit dubious, because `i8x8` does not currently need to map to an MMX register, but `__m64` does. 

Maybe the current approach of hardcoding `#[repr(simd)] struct __m64(i64)` to map to `x86_mmx` is wrong and we should be mapping all 64-bit `#[repr(simd)]` to `x86_mmx` instead. Some intrinsics take `__m64` but we would like to strongly type them as `i16x4` but currently can't (if we want proper code gen) due to that mismatch, so we need to use `__m64` instead.

Reply:
It doesn't matter, this is just mapping to a 64-bit vector type.

Reply:
So IIUC it is possible to change all the intrinsics that now take an `i8x8` to use an `i16x4` or `__m64` or any type with 64-bit wide argument right?

Reply:
Yes just like with `__m256i` you can have any `iAxB` so long as `A * B = 256`. In general I don't expect this to be a problem as well probaly hardcode everything to `__m256i` anyway.